### PR TITLE
Adding snapshot module

### PR DESCRIPTION
**Pull request by zikalino**
##### SUMMARY
Adding snapshot module to be used by the gallery

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME
azure_rm_snapshot

### CODE REVIEW

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+short_description: Manage Azure Snapshot instance.
+description:
+  - 'Create, update and delete instance of Azure Snapshot.'
+options:
+  resource_group:
+    description:
+      - The name of the resource group.
+    required: true
+    type: str
+  name:
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_snapshot.py:29]: is this not required?

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+      - Resource name
+    type: str
+  location:
+    description:
+      - Resource location
+    required: true
+    type: str
+  sku:
+    description:
+      - undefined
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_snapshot.py]: ?

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+    required: true
+    type: str
+  name:
+    description:
+      - Resource name
+    type: str
+  location:
+    description:
+      - Resource location
+    required: true
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_snapshot.py]: is location always required even when updating?

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+    type: str
+  sku:
+    description:
+      - undefined
+    type: dict
+    suboptions:
+      name:
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_snapshot.py:42]: required?

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+        description:
+          - The sku name.
+        type: str
+      tier:
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_snapshot.py:50]: required?

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+        description:
+          - The sku tier.
+        type: str
+  os_type:
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_snapshot.py:54]: choices?

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+    description:
+      - The Operating System type.
+    type: str
+  creation_data:
+    description:
+      - >-
+        Disk source information. CreationData information cannot be changed
+        after the disk has been created.
+    required: true
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_snapshot.py]: always required even when updating?

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+    type: dict
+    suboptions:
+      create_option:
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_snapshot.py:68]: choices?

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+                disposition='/',
+                required=True
+            ),
+            sku=dict(
+                type='dict',
+                disposition='/',
+                options=dict(
+                    name=dict(
+                        type='str',
+                        choices=['Standard_LRS',
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_snapshot.py:163]: put this into doc

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+                                 'Premium_LRS',
+                                 'Standard_ZRS']
+                    )
+                )
+            ),
+            os_type=dict(
+                type='str',
+                disposition='/properties/osType',
+                choices=['Windows',
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_snapshot.py:175]: put this into doc

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+            ),
+            creation_data=dict(
+                type='dict',
+                disposition='/properties/creationData',
+                required=True,
+                options=dict(
+                    create_option=dict(
+                        type='str',
+                        disposition='createOption',
+                        choices=['Empty',
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_snapshot.py]: add to doc

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py [hunk 12]
```diff
@@ -0,0 +1,378 @@
...
+                setattr(self, key, kwargs[key])
+            elif kwargs[key] is not None:
+                self.body[key] = kwargs[key]
+
+        self.inflate_parameters(self.module_arg_spec, self.body, 0)
+
+        old_response = None
+        response = None
+
+        self.mgmt_client = self.get_mgmt_svc_client(GenericRestClient,
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_snapshot.py:237]: why use rest client? why not using sdk?

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py [hunk 13]
```diff
@@ -0,0 +1,378 @@
...
+        if response:
+           self.results["id"] = response["id"]
+
+        return self.results
+
+    def create_update_resource(self):
+        # self.log('Creating / Updating the Snapshot instance {0}'.format(self.))
+
+        try:
+            response = self.mgmt_client.query(self.url,
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_snapshot.py]: not using  positional paramter, pls specify argument names.

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py [hunk 14]
```diff
@@ -0,0 +1,378 @@
...
+        except Exception:
+            response = {'text': response.text}
+            pass
+
+        return response
+
+    def delete_resource(self):
+        # self.log('Deleting the Snapshot instance {0}'.format(self.))
+        try:
+            response = self.mgmt_client.query(self.url,
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_snapshot.py]: not using positional argument list, pls specify argument names

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py [hunk 15]
```diff
@@ -0,0 +1,378 @@
...
+            self.log('Error attempting to delete the Snapshot instance.')
+            self.fail('Error deleting the Snapshot instance: {0}'.format(str(e)))
+
+        return True
+
+    def get_resource(self):
+        # self.log('Checking if the Snapshot instance {0} is present'.format(self.))
+        found = False
+        try:
+            response = self.mgmt_client.query(self.url,
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_snapshot.py]: not using positional argument list, pls specify argument names

**Review by yungezz — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py (hunk 1 above)
Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_snapshot.py:29]: name is required

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py (hunk 2 above)
Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_snapshot.py]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py (hunk 3 above)
Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_snapshot.py]: fixed, not required

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py (hunk 4 above)
Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_snapshot.py:42]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py (hunk 5 above)
Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_snapshot.py:50]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py (hunk 6 above)
Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_snapshot.py:54]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py (hunk 7 above)
Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_snapshot.py]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py (hunk 8 above)
Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_snapshot.py:68]: fixed

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py (hunk 9 above)
Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_snapshot.py:163]: done

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py (hunk 11 above)
Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_snapshot.py]: fixed, and added Import as default and only option

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py (hunk 12 above)
Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_snapshot.py:237]: it was faster to do without running into dependency problems.

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py (hunk 1 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_snapshot.py:29]: then pls set "required: true" in doc

File: lib/ansible/modules/cloud/azure/azure_rm_snapshot.py (hunk 5 above)
Comment (yungezz) [reply lib/ansible/modules/cloud/azure/azure_rm_snapshot.py:50]: no required, but only choices

**Review by yungezz — APPROVED** [review]

**Pull request merged**